### ISSUE TITLE
5.8.2 segfaults on s390x

### ISSUE THREAD

**Issue by jamesjer** [polyml/polyml#150]
I tried building the 5.8.2 release for Fedora.  The build failed on the s390x builder.  That is a big endian architecture.  The end of the build looks like this:
```
make[2]: Entering directory '/builddir/build/BUILD/polyml-5.8.2'
./polyimport  polytemp.txt -I . < ./exportPoly.sml
Use: basis/build.sml
Use: basis/InitialBasis.ML
make[2]: *** [Makefile:1168: polyexport.o] Error 139
make[2]: Leaving directory '/builddir/build/BUILD/polyml-5.8.2'
make[1]: *** [Makefile:726: all-recursive] Error 1
make: *** [Makefile:485: all] Error 2
```

Running under gdb:
```
Use: basis/build.sml
Use: basis/InitialBasis.ML

Thread 1 "polyimport" received signal SIGSEGV, Segmentation fault.
PolyObject::GetConstSegmentForCode (count=<synthetic pointer>: <optimized out>,
    cp=<synthetic pointer>: 0xe90003fffdfea028, obj_length=<optimized out>, this=0xe90003fffdfea028) at globals.h:341
341                count = cp[-1].AsUnsigned();
(gdb) bt
#0  PolyObject::GetConstSegmentForCode (count=<synthetic pointer>: <optimized out>,
    cp=<synthetic pointer>: 0xe90003fffdfea028, obj_length=<optimized out>, this=0xe90003fffdfea028) at globals.h:341
#1  ScanAddress::ScanAddressesInObject (this=0x3fffffff038, obj=0xe90003fffdfea028, lengthWord=<optimized out>)
    at scanaddrs.cpp:85
#2  0x000003fffded49c0 in ScanAddress::ScanAddressesInObject (base=<optimized out>, this=0x3fffffff038)
    at scanaddrs.h:90
#3  ScanAddress::ScanAddressesInRegion (this=0x3fffffff038, region=<optimized out>, end=0x3fffdfeb000)
    at scanaddrs.cpp:195
#4  0x000003fffdecc4e8 in RunQuickGC (wordsRequiredToAllocate=<optimized out>) at quick_gc.cpp:597
#5  RunQuickGC (wordsRequiredToAllocate=<optimized out>) at quick_gc.cpp:534
#6  0x000003fffdeb4a0c in QuickGCRequest::Perform (this=0x3fff9c7ac50) at gc.cpp:377
#7  0x000003fffdecd4a2 in Processes::BeginRootThread (this=0x3fffdf0c970 <_ZL15processesModule.lto_priv.0>,
    rootFunction=<optimized out>) at processes.cpp:1432
#8  0x000003fffdec1110 in polymain (argc=<optimized out>, argv=<optimized out>, exports=<optimized out>)
    at mpoly.cpp:412
#9  0x000003fffdc2bdf4 in __libc_start_main (main=0x10004f0 <main>, argc=<optimized out>, argv=0x3fffffff4f8,
    init=<optimized out>, fini=<optimized out>, rtld_fini=0x3fffdf90590 <_dl_fini>, stack_end=0x3fffffff440)
    at libc-start.c:332
#10 0x000000000100053e in _start ()
(gdb) print *cp
Cannot access memory at address 0xe90003fffdfea028
```

Running under valgrind:
```
==133484== Memcheck, a memory error detector
==133484== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==133484== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==133484== Command: .libs/polyimport polytemp.txt -I .
==133484==
Use: basis/build.sml
Use: basis/InitialBasis.ML
==133484== Invalid read of size 8
==133484==    at 0x489B798: UnknownInlinedFun (globals.h:183)
==133484==    by 0x489B798: UnknownInlinedFun (globals.h:341)
==133484==    by 0x489B798: ScanAddress::ScanAddressesInObject(PolyObject*, unsigned long) (scanaddrs.cpp:85)
==133484==    by 0x48A19BF: UnknownInlinedFun (scanaddrs.h:90)
==133484==    by 0x48A19BF: ScanAddress::ScanAddressesInRegion(PolyWord*, PolyWord*) (scanaddrs.cpp:195)
==133484==    by 0x48994E7: UnknownInlinedFun (quick_gc.cpp:597)
==133484==    by 0x48994E7: RunQuickGC(unsigned long) (quick_gc.cpp:534)
==133484==    by 0x4881A0B: QuickGCRequest::Perform() (gc.cpp:377)
==133484==    by 0x489A4A1: Processes::BeginRootThread(PolyObject*) (processes.cpp:1432)
==133484==    by 0x488E10F: polymain (mpoly.cpp:412)
==133484==    by 0x4913DF3: (below main) (libc-start.c:332)
==133484==  Address 0xe900000004028020 is not stack'd, malloc'd or (recently) free'd
==133484==
==133484==
==133484== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==133484==  Access not within mapped region at address 0xE900000004028000
==133484==    at 0x489B798: UnknownInlinedFun (globals.h:183)
==133484==    by 0x489B798: UnknownInlinedFun (globals.h:341)
==133484==    by 0x489B798: ScanAddress::ScanAddressesInObject(PolyObject*, unsigned long) (scanaddrs.cpp:85)
==133484==    by 0x48A19BF: UnknownInlinedFun (scanaddrs.h:90)
==133484==    by 0x48A19BF: ScanAddress::ScanAddressesInRegion(PolyWord*, PolyWord*) (scanaddrs.cpp:195)
==133484==    by 0x48994E7: UnknownInlinedFun (quick_gc.cpp:597)
==133484==    by 0x48994E7: RunQuickGC(unsigned long) (quick_gc.cpp:534)
==133484==    by 0x4881A0B: QuickGCRequest::Perform() (gc.cpp:377)
==133484==    by 0x489A4A1: Processes::BeginRootThread(PolyObject*) (processes.cpp:1432)
==133484==    by 0x488E10F: polymain (mpoly.cpp:412)
==133484==    by 0x4913DF3: (below main) (libc-start.c:332)
==133484==  If you believe this happened as a result of a stack
==133484==  overflow in your program's main thread (unlikely but
==133484==  possible), you can try to increase the size of the
==133484==  main thread stack using the --main-stacksize= flag.
==133484==  The main thread stack size used in this run was 8388608.
==133484==
==133484== HEAP SUMMARY:
==133484==     in use at exit: 248,512 bytes in 65 blocks
==133484==   total heap usage: 96 allocs, 31 frees, 522,530 bytes allocated
==133484==
==133484== LEAK SUMMARY:
==133484==    definitely lost: 0 bytes in 0 blocks
==133484==    indirectly lost: 0 bytes in 0 blocks
==133484==      possibly lost: 1,728 bytes in 6 blocks
==133484==    still reachable: 246,784 bytes in 59 blocks
==133484==                       of which reachable via heuristic:
==133484==                         newarray           : 73,040 bytes in 1 blocks
==133484==         suppressed: 0 bytes in 0 blocks
==133484== Rerun with --leak-check=full to see details of leaked memory
==133484==
==133484== For lists of detected and suppressed errors, rerun with: -s
==133484== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
Segmentation fault
```

Version 5.8.1 builds successfully on s390x.  I would be grateful for any suggestions on where to look for the source of the problem.

**Comment by dcjm** [comment]
Thanks for reporting this. It was an endian issue.  I've pushed a fix for it and rebuilt the 64-bit interpreted pre-built compiler.  It has been tested in an emulator but not on the real hardware. This also affects 32-bit big-endian so the 32-bit interpreted compiler also needs to be rebuilt.

**Comment by jamesjer** [comment]
Thank you for the quick fix.  Building from the fixes-5.8.2 branch worked on s390x.
